federation/pkg/federation-controller/service/ingress/ingress.go

```diff
@@ -0,0 +1,152 @@
+/*
+Copyright 2016 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package ingress
+
+import (
+	"encoding/json"
+	"strings"
+
+	fed "k8s.io/kubernetes/federation/apis/federation"
+	"k8s.io/kubernetes/pkg/api/v1"
+)
+
+const (
+	FederationServiceIngressAnnotation = "federation.kubernetes.io/service-ingress-endpoints"
+)
+
+type serviceIngress fed.FederatedServiceIngress
+
+func NewServiceIngress() *serviceIngress {
+	return &serviceIngress{}
+}
+
+func (si *serviceIngress) AddEndpoints(region, zone, cluster string, endpoints []string, healthy bool) *serviceIngress {
+	if si.Endpoints == nil {
+		si.Endpoints = make(map[string]map[string]map[string]*fed.ClusterServiceIngress)
+	}
+	if si.Endpoints[region] == nil {
+		si.Endpoints[region] = make(map[string]map[string]*fed.ClusterServiceIngress)
+	}
+	if si.Endpoints[region][zone] == nil {
+		si.Endpoints[region][zone] = make(map[string]*fed.ClusterServiceIngress)
+	}
+	clusterIngEps, ok := si.Endpoints[region][zone][cluster]
+	if !ok {
+		clusterIngEps = &fed.ClusterServiceIngress{}
+	}
+	clusterIngEps.Endpoints = append(clusterIngEps.Endpoints, endpoints...)
+	clusterIngEps.Healthy = healthy
+	si.Endpoints[region][zone][cluster] = clusterIngEps
+	return si
+}
+
+func (si *serviceIngress) RemoveEndpoint(region, zone, cluster string, endpoint string) *serviceIngress {
+	clusterIngress := si.Endpoints[region][zone][cluster]
+	for i, ep := range clusterIngress.Endpoints {
+		if ep == endpoint {
+			clusterIngress.Endpoints = append(clusterIngress.Endpoints[:i], clusterIngress.Endpoints[i+1:]...)
+		}
+	}
+	if len(clusterIngress.Endpoints) < 1 {
+		clusterIngress.Healthy = false
```

Comment:
There are 2 things which we represent in clusterIngress, one is the external load-balancer ip(or host) of the service, other one is whether that external load-balancer has backing pods within the cluster (or so to say service endpoints exists) to handle the data traffic. These 2 piece of information could be segregated to act upon them independently for making the implementation simple. Say for scenario, the external load-balancer is created but there are no backing pods yet, we would add the Ip and set Healthy field to false and after a while the backing pods are created. at this point all we would need to do is to update Healthy from false to true instead of trying to fetch the service load-balancer ip and update.
Also i tend to agree with you this field is not needed absolutely. and its just a implementation preference. I would do away with this field, which would make the Annotation in service to be simple.